### DISCUSSION TITLE
How to specify the constant type（int32， int64...） in an SQL expression； as if the default is int64， how to specify int32

### DISCUSSION THREAD

**Discussion by hrh007** [apache/datafusion#6583] [Q&A]
when in write sql select column1 = 123, and column1 type is int32, it doesn't work;     

log is :   Error: Internal("The type of Int32 = Int64 of binary physical should be same")

**Comment by alamb** [comment]
Hi @hrh007  -- that definitely sounds like a bug - if you have a reproducer I would love to see it

A workaround might be to put an explicit cast in:

For example this would cast `123` to bigint (Int64):
```sql
select column1 = 123::bigint
```

The SQL --> Arrow data type mapping for DataFusion is on https://arrow.apache.org/datafusion/user-guide/sql/data_types.html

**Reply by hrh007** [reply]
you can use SQL for example:  select  arrow_typeof(123);   the result is int64

**Reply by hrh007** [reply]
Another example: register_parquet,  and column1 type is int32
select column1 = 123
The SQL statement reports an error： Error: Internal("The type of Int32 = Int64 of binary physical should be same")

**Reply by hrh007** [reply]
And if use typecast, SQL is : select udf_xxx( column1 = 3015::int, ...)

Error: NotImplemented("Physical plan does not support logical expression; just used by udf udf_xxx( table5.column1 = Int32(3015), ) AS udtf_xxx(table5.column1 = Int64(3015))")

Physical plan and  logical expression type mismatch

**Reply by alamb** [reply]
it sounds like a bug to me -- perhaps you can provide a reproducer so we can investigate it more carefully?

**Reply by hrh007** [reply]
Can my examples above be reproduced in your environment ？

**Reply by alamb** [reply]
I tried to reproduce it using `datafusion-cli` and I could not:

```sql
DataFusion CLI v26.0.0
❯ create table t as select 1::int as column1;
0 rows in set. Query took 0.031 seconds.
❯ select * from t;
+---------+
| column1 |
+---------+
| 1       |
+---------+
1 row in set. Query took 0.005 seconds.
❯ select arrow_typeof(column1) from t;
+-------------------------+
| arrow_typeof(t.column1) |
+-------------------------+
| Int32                   |
+-------------------------+
❯ select column1 = 123 from t;
+------------------------+
| t.column1 = Int64(123) |
+------------------------+
| false                  |
+------------------------+
1 row in set. Query took 0.005 seconds.

```

**Reply by hrh007** [reply]
My version is datafusion 14.0
Your example above is also correct in my cli env； 
but when I write an rust example in datafusion source code : register_parquet, run a sql; the wrong will occur
